core/src/main/java/org/jclouds/http/internal/BaseHttpCommandExecutorService.java

```diff
@@ -80,6 +81,64 @@ protected BaseHttpCommandExecutorService(HttpUtils utils, ContentMetadataCodec c
       this.wire = checkNotNull(wire, "wire");
       this.idempotentMethods = ImmutableSet.copyOf(idempotentMethods.split(","));
    }
+   @Override
+   public ListenableFuture<HttpResponse> invokeAsync(final HttpCommand command){
+      SettableFuture<HttpResponse> finalFuture = SettableFuture.create();
+      this.invokeAsyncInternal(command, finalFuture);
+      return finalFuture;
+   }
+
+   private void invokeAsyncInternal(final HttpCommand command, final SettableFuture<HttpResponse> finalFuture) {
+      ListenableFuture<HttpResponse> future = null;
+      try {
+         final HttpRequest request = command.getCurrentRequest();
+         final Q nativeRequest = getNativeRequest(request);
+         future = invokeAsync(nativeRequest);
+         Futures.addCallback(future, new FutureCallback<HttpResponse>() {
+            @Override
+            public void onSuccess(HttpResponse response) {
+               logResponse(request, response);
+               int statusCode = response.getStatusCode();
+               if (statusCode >= 300) {
+                  if (shouldContinue(command, response))
+                     invokeAsyncInternal(command, finalFuture);
+                  else
+                     cleanup(nativeRequest);
+                  setResultOrException(command, finalFuture, response);
+               } else {
+                  cleanup(nativeRequest);
+                  setResultOrException(command, finalFuture, response);
+               }
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+               handleInvokeAsyncFailure(t, command, finalFuture);
+            }
+         });
+      } catch (Exception e) {
+         handleInvokeAsyncFailure(e, command, finalFuture);
+
+      }
```

Comment:
>Are you suggesting to change the return type of invoke method to Future?

Yes. It could not look OK, but it is an internal method, and I wanted to explore if we could benefit from having just one way of invoking the underlying transport method. I just wanted to see if the overhead introduced by the immediate future are worth having a cleaner implementation with just one method.